### PR TITLE
Make sure the build.rustc version is either the same or 1 apart

### LINKED ISSUE #110067: Confusing diagnostics for unstable features when building Rust with stable rustc
I'm trying to build Rust on a Tier 3 platform (`riscv64-alpine-linux-musl`; actually it's `riscv64-unknown-musl` that's listed in Tier 3, but I don't think that affects the problem). Since there are no snapshots for this platform, I installed rustc from the distro's package manager (apk) and created a `config.toml` file to force using the pre-installed rustc rather than downloading snapshots:

```
profile = "compiler"

[build]
rustc  = "/usr/bin/rustc"
cargo  = "/usr/bin/cargo"
rustfmt = "/usr/bin/rustfmt"
host   = ["riscv64-alpine-linux-musl"]
target = ["riscv64-alpine-linux-musl"]

[rust]
musl-root = "/usr"
debug     = true

[target.riscv64-alpine-linux-musl]
musl-libdir = "/usr/lib"
```

And the output of `rustc --version -v`:

```
# rustc --version -v
rustc 1.68.2 (9eb3afe9e 2023-03-27) (Alpine Linux)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: riscv64-alpine-linux-musl
release: 1.68.2
LLVM version: 15.0.7
```

I ran `./x.py build`; when it got as far as building the core library, I got a series of errors that were very hard to make sense of until I realized that it's because core library code depends on unstable features and I was using a stable rustc to compile it:

```
building stage0 library artifacts (riscv64-alpine-linux-musl) 
  Downloaded adler v1.0.2
  Downloaded miniz_oxide v0.5.3
  Downloaded unicode-width v0.1.10
  Downloaded hashbrown v0.12.3
  Downloaded compiler_builtins v0.1.91
  Downloaded cc v1.0.77
  Downloaded addr2line v0.17.0
  Downloaded rustc-demangle v0.1.21
  Downloaded gimli v0.26.2
  Downloaded 9 crates (1.2 MB) in 1.45s
   Compiling cc v1.0.77
   Compiling compiler_builtins v0.1.91
   Compiling core v0.0.0 (/home/scratch/rust/library/core)
error: expected identifier, found keyword `move`
   --> library/core/src/ops/try_trait.rs:395:15
    |
395 |         const move |a, b| NeverShortCircuit(f(a, b))
    |               ^^^^ expected identifier, found keyword

error: expected one of `:`, `;`, or `=`, found `|`
   --> library/core/src/ops/try_trait.rs:395:20
    |
395 |         const move |a, b| NeverShortCircuit(f(a, b))
    |                    ^ expected one of `:`, `;`, or `=`
```

There were many more errors, but I think they all relate to unstable features.

It took me ages to figure out that (for example) the first error is due to the `const_closures` feature being disabled in stable.

I understand that the stable compiler won't allow using unstable features even with `-Zunstable-options`, but is it possible for the parser to recognize this code even so and reject it with something like "feature `const_closures` is not enabled" instead of the much more generic "expected identifier" error?

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #110067
r? @jyn514

### CODE REVIEW

**Comment by jyn514** [comment]
```
failed to execute command: "/checkout/obj/build/aarch64-apple-darwin/stage0/bin/rustc" "--version"
error: Exec format error (os error 8)

error: test failed, to rerun pass `--lib`

thread 'config::tests::download_ci_llvm' panicked at 'status code: 1', lib.rs:1722:9
```

I think you need to slap if !builder.dry_run() on this new check.

This looks good! r=me if my suggestion fixes the build :)

**Comment by chenyukang** [comment]
> failed to execute command: "/checkout/obj/build/aarch64-apple-darwin/stage0/bin/rustc" "--version"

it turns out we are trying to run a binary of `aarch64-apple-darwin` on an `x86_64-gnu` VM, I assume this change is safe for the tests:
https://github.com/rust-lang/rust/pull/111538/commits/3e41587e62d0ba17e4daa64d0f8d9261b8abdc5a

seems those assertions are verifying logic hard-coded in Rust.

**Comment by jyn514** [comment]
yup that's totally fine! thanks for looking into this :)

@bors delegate=chenyukang (use r=jyn514)

File: src/bootstrap/config.rs [hunk 1]
```diff
@@ -1680,6 +1682,41 @@ impl Config {
...
+                .unwrap();
+        if !(source_version == rustc_version
+            || (source_version.major == rustc_version.major
+                && source_version.minor == rustc_version.minor + 1))
+        {
+            let prev_version = format!("{}.{}.x", source_version.major, source_version.minor - 1);
+            panic!(
+                "Unexpected rustc version: {}, we should use {}/{} to build source with {}",
+                rustc_version, prev_version, source_version, source_version
+            );
```
Comment (bjorn3) [inline src/bootstrap/config.rs:1716]: I think this should use `eprintln!()` followed by `crate::detail_exit(1)`. Panicking will show a useless backtrace when `RUST_BACKTRACE=1` is used.

**Pull request merged**